### PR TITLE
Gh 1060 multi property aggregation

### PR DESCRIPTION
**Pull request by p013570**
A lot of the files that have changed are simply the schema json files.

dataSchema.json -> elements.json
dataTypes.json + storeTypes.json -> types.json

In some cases we used a storeSchema.json file that contained a vertex serialiser. This has been renamed to serialisation.json.

Advanced aggregation and validation has been separated out into aggregation.json and validation.json. 

Renaming these schema files is not a breaking change and does not require users to do the same. I just think that for new users it will be clearer and easier to understand if we split the schema into just the elements and the types.

I have written quite a few unit tests to check the logic for the multi property schema aggregation, but it is a bit tricky so please review that bit carefully.

The change has added an extra section to the Schema but this is optional so hopefully it won't complicate things for novice users of Gaffer. There was a bit of complexity around the change due to not be able to aggregation multiple properties together if they are stored in different places. For example - we will not allow groupBy and non-groupBy properties to be aggregated using the same BinaryOperator. 

See #1060 for a discussion around this feature and for the update Schema wiki page content.

Before we merge this in we should check that we are all happy with the extra "aggregator" section in the schema and that it doesn't complicate things for users.

### CODE REVIEW

**Comment by p013570** [comment]
Closing whilst I make a few changes.

File: core/common-util/src/main/java/uk/gov/gchq/gaffer/commonutil/CollectionUtil.java [hunk 1]
```diff
@@ -46,4 +47,36 @@ private CollectionUtil() {
 
         return treeSet;
     }
+
+    public static boolean containsAny(final Collection collection, final Object[] objects) {
```
Comment (t616178) [inline core/common-util/src/main/java/uk/gov/gchq/gaffer/commonutil/CollectionUtil.java]: Is it a good idea to write these ourselves? There are several other solutions from the libraries:

```
Collections.disjoint(collection, objects); // pure Java libraries
CollectionUtils.containsAny(collection, object); // Apache Commons
```

File: core/data/src/main/java/uk/gov/gchq/gaffer/data/element/function/ElementAggregator.java [hunk 2]
```diff
@@ -57,6 +60,19 @@ public Properties apply(final Properties state, final Properties properties) {
...
     @Override
+    public List<TupleAdaptedBinaryOperator<String, ?>> getComponents() {
+        if (readOnly) {
+            return Collections.unmodifiableList(super.getComponents());
+        }
+
+        return super.getComponents();
+    }
+
+    public void lock() {
```
Comment (t616178) [inline core/data/src/main/java/uk/gov/gchq/gaffer/data/element/function/ElementAggregator.java]: Do we need to be able to unlock as well?

File: core/store/src/main/java/uk/gov/gchq/gaffer/store/schema/SchemaElementDefinition.java [hunk 3]
```diff
@@ -474,6 +583,20 @@ public CHILD_CLASS identifiers(final Map<IdentifierType, String> identifiers) {
...
         }
 
+        public CHILD_CLASS aggregator(final ElementAggregator aggregator) {
+            elDef.aggregator = aggregator;
+            return self();
+        }
+
+        @JsonSetter("aggregateFunctions")
+        public CHILD_CLASS aggregateFunctions(final List<TupleAdaptedBinaryOperator<String, Tuple<String>>> aggregateFunctions) {
+            if (null == getElementDef().aggregator) {
```
Comment (t616178) [inline core/store/src/main/java/uk/gov/gchq/gaffer/store/schema/SchemaElementDefinition.java]: Is this direct access to the property definitely safe?

File: core/data/src/main/java/uk/gov/gchq/gaffer/data/element/function/ElementAggregator.java (hunk 2 above)
Comment (p013570) [inline core/data/src/main/java/uk/gov/gchq/gaffer/data/element/function/ElementAggregator.java]: No, I think once the ElementAggregator is locked - no changes should be allowed. If you need to make changes then you should create a new instance of ElementAggregator.

File: core/common-util/src/main/java/uk/gov/gchq/gaffer/commonutil/CollectionUtil.java (hunk 1 above)
Comment (p013570) [inline core/common-util/src/main/java/uk/gov/gchq/gaffer/commonutil/CollectionUtil.java]: Those 2 methods takes collection, collection. We need to ones that works with an array. I could do Arrays.asList(array) and then use those methods but it seemed wasteful to create an extra object. What do you think?

Comment (t616178) [inline core/common-util/src/main/java/uk/gov/gchq/gaffer/commonutil/CollectionUtil.java]: That sounds reasonable - I just wanted to make sure we'd considered use of the libraries.

File: core/data/src/main/java/uk/gov/gchq/gaffer/data/element/function/ElementAggregator.java (hunk 2 above)
Comment (t616178) [inline core/data/src/main/java/uk/gov/gchq/gaffer/data/element/function/ElementAggregator.java]: Fine.

**Pull request merged**